src/core/utils/AlbedoNormalPass.js

```diff
@@ -0,0 +1,224 @@
+						return;
+					
+						vec3 mapN = texture(normalMap, vUv).xyz * 2.0 - 1.0;
+						mapN.xy *= normalScale;
+						mat3 tbn = mat3(normalize(vTangent), normalize(vBitangent), normalize(vNormal));
+						vec3 worldNormal = normalize(tbn * mapN);
+						
+						if (useWorldSpaceNormals) {
+							normal = worldNormal;
+						} else {
+							normal = normalize((viewMatrix * vec4(worldNormal, 0.0)).xyz);
+						}
+					} else {
+						normal = useWorldSpaceNormals ? vWorldNormal : vNormal;
+					}
+
+					normal = normalize(normal + vec3(1e-6));
+					normal = normal * 0.5 + 0.5;
+					gNormal = vec4(normal, 1.0);
+                }
+            `,
+			glslVersion: THREE.GLSL3
+		} );
+
+	}
+
+	async render( renderer, scene, camera ) {
+
+		const target = this.renderTarget;
+		target.setSize( renderer.domElement.width, renderer.domElement.height );
+
+		this.swapMaterials( scene );
+
+		const oldRenderTarget = renderer.getRenderTarget();
+		renderer.setRenderTarget( target );
+		renderer.render( scene, camera );
+		renderer.setRenderTarget( oldRenderTarget );
+
+		this.restoreMaterials();
+
+		// return the two textures
+		return { albedo: target.textures[ 0 ], normal: target.textures[ 1 ] };
+
+	}
+
+	swapMaterials( object ) {
+
+		if ( object instanceof THREE.Mesh && object.material ) {
+
+			if ( ! this.originalMaterials.has( object ) )
+				this.originalMaterials.set( object, object.material );
+
+			if ( this.albedoMaterials.has( object ) ) {
+
+				object.material = this.albedoMaterials.get( object );
+				return;
+
+			}
+
+			const material = object.material;
+			const newAlbedoMaterial = this.albedoNormalMaterial.clone();
+
+			if ( material.color ) newAlbedoMaterial.uniforms.color.value.copy( material.color );
+			newAlbedoMaterial.uniforms.diffuseMap.value = material.map;
+			newAlbedoMaterial.uniforms.normalMap.value = material.normalMap;
+			newAlbedoMaterial.uniforms.useDiffuseMap.value = !! material.map;
+			newAlbedoMaterial.uniforms.useNormalMap.value = !! material.normalMap;
+			newAlbedoMaterial.uniforms.debugNormalMaps.value = this.debugNormalMaps;
+
+			if ( material.normalMap ) {
+
+				newAlbedoMaterial.defines.USE_TANGENT = '';
+
+			}
+
+			this.albedoMaterials.set( object, newAlbedoMaterial );
+			object.material = newAlbedoMaterial;
+
+		}
+
+		// biome-ignore lint/complexity/noForEach: <explanation>
+		object.children.forEach( child => this.swapMaterials( child ) );
+
+	}
+
+	restoreMaterials() {
+
+		this.originalMaterials.forEach( ( material, object ) => {
+
+			if ( object instanceof THREE.Mesh ) {
+
+				object.material = material;
+
+			}
+
+		} );
+		this.originalMaterials.clear();
+
+	}
+
+	setUseWorldSpaceNormals( value ) {
+
+		this.useWorldSpaceNormals = value;
+		// biome-ignore lint/complexity/noForEach: <explanation>
+		this.albedoMaterials.forEach( ( material ) => {
+
+			if ( material instanceof THREE.ShaderMaterial ) {
+
+				material.uniforms.useWorldSpaceNormals.value = value;
+
+			}
+
+		} );
```

Comment:
Is there a reason to not just use one kind of normals? I'd like to avoid a toggle if possible. We should just pick the best one of there is one.

Reply:
Agreed, OIDN said it supports both and I didn't want to force the decision and wasn't sure if one was better than the other (especially considering normal maps) . Whichever way you want to go is fine.

Reply:
I think using camera-space normals is the easiest to understand.